### ISSUE TITLE
Eslint 9 (2024)

### ISSUE THREAD

**Issue by lb-** [wagtail/eslint-config-wagtail#22]
Just adding an issue to help us track the upcoming Eslint 9.0 release. 

A few key changes will come, including dropping support for all node versions under 20. 

https://eslint.org/blog/2023/11/whats-coming-in-eslint-9.0.0/#dropping-support-for-node.js-%3C-v18.18.0%2C-v19

**Comment by thibaudcolas** [comment]
This will be a good occasion for us to adopt the new config system in particular!

**Comment by lb-** [comment]
I just gave this a go locally, but I think we will be blocked until AirBNB update their package. 
https://github.com/airbnb/javascript/issues/2961

Unless we break away from this, but that could be a lot of extra work.

https://github.com/lb-/eslint-config-wagtail/tree/feature/22-upgrade-eslint-v9 (rough start, not functional)

**Comment by thibaudcolas** [comment]
@lb- I wonder if we should break away from it. ESLint’s built-in rules, and Airbnb’s interpretation of the rules, and our overrides on top – don’t change particularly often?

And I think Airbnb’s config has lost its relevance for us a bit:

- We don’t do as much React as we used to.
- We do a lot of Stimulus, for which there are relevant vanilla JS rules, but not many
- We do more and more TypeScript, which isn’t part of Airbnb’s coding style

@laymonage what do you think?

**Comment by lb-** [comment]
I'm all for us to move away from Airbnb's package and would be happy to try to give this a go. I'll wait on Sage's thoughts though. 

I feel it would give us a chance to simplify sibling dependencies also.

**Comment by laymonage** [comment]
I agree with Thibaud points, I think we're better off dropping AirBnB's config. It might even make it easier if we want to migrate to other tooling in the future (e.g. Biome).

**Comment by laymonage** [comment]
I tried using Copilot to do this for us in #33 while keeping some AirBnB rules that might still be useful, but I'm not sure I want to review every single rule in that PR.

I think I'll try to start over with just using the recommended rules from `eslint`, `eslint-plugin-import`, `eslint-plugin-jsx-a11y`, `eslint-plugin-react`, `eslint-plugin-react-hooks`, and `eslint-config-prettier`. And then from there, we review if there's any rule we'd like to override to be made more loose/strict.

**Comment by laymonage** [comment]
Done in #36, pending a new release 🎉 

ESLint v10 has been released but typescript-eslint doesn't support it yet: https://github.com/typescript-eslint/typescript-eslint/issues/11952
